### PR TITLE
Address Issue #807 (Only use a single item and center it for a double chest buy/sell shop.)

### LINKED ISSUE #807: [Feature] Only use a single item and center it for a double chest buy/sell shop.
**Describe the Feature**
Only use a single item and center it for a double chest buy/sell shop. Primarily to prevent clutter both with floating items and nametag display. There's really no reason to display twice when the shop buys and sells the same item & shares a storage together.

**Screenshots**
link

### PR DESCRIPTION
**Pull request by Starmism**
Hi! This is my first contribution to this project, so any help would be appreciated in getting my code merged!

This PR resolves #807 
First, it checks for any situations where there is a double chest shop of the same item.
Then it finds the left block of that double chest (in relation to the chest's direction), and tells it not to spawn an item.
The right chest then spawns it's item centered between the two chests.

This removes the duplicate items of the same type for a double chest, especially useful when you want to buy and sell the same item from the same chest.

### CODE REVIEW

File: src/main/java/org/maxgamer/quickshop/shop/DisplayItem.java [hunk 1]
```diff
@@ -136,6 +136,7 @@ public static boolean checkIsTargetShopDisplay(@NotNull final ItemStack itemStac
         }
         String defaultMark = ShopProtectionFlag.getDefaultMark();
         String shopLocation = shop.getLocation().toString();
+        String attachedShopLocation = (!shop.isRealDouble()) ? "null" : (shop.getAttachedShop() == null ? "null" : shop.getAttachedShop().getLocation().toString());
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/DisplayItem.java:139]: Please convert to an if statement, ternary expression will make it unreadable

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 2]
```diff
@@ -387,6 +413,7 @@ public void delete() {
     @Override
     public void delete(boolean memoryOnly) {
         Util.ensureThread(false);
+        ContainerShop neighbor = getAttachedShop();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:416]: Emmmm, I think we should put the variable to where we actually need to use as nearly as possible.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 3]
```diff
@@ -401,19 +428,30 @@ public void delete(boolean memoryOnly) {
         if (memoryOnly) {
             // Delete it from memory
             plugin.getShopManager().removeShop(this);
+            onUnload();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:431]: Why we need to call `onUnload()` for double times?

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 4]
```diff
@@ -401,19 +428,30 @@ public void delete(boolean memoryOnly) {
...
         } else {
             // Delete the signs around it
             for (Sign s : this.getSigns()) {
                 s.getBlock().setType(Material.AIR);
             }
+            if (displayItem != null) {
+                displayItem.remove();
+            }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:439]: `onUnload()` method will do it

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 5]
```diff
@@ -401,19 +428,30 @@ public void delete(boolean memoryOnly) {
...
             // Delete it from the database
             // Refund if necessary
             if (plugin.getConfig().getBoolean("shop.refund")) {
-                plugin.getEconomy().deposit(this.getOwner(), plugin.getConfig().getDouble("shop.cost"), getLocation().getWorld(), getCurrency());
+                plugin.getEconomy()
+                    .deposit(this.getOwner(), plugin.getConfig().getDouble("shop.cost"),
+                        Objects.requireNonNull(getLocation().getWorld()), getCurrency());
             }
             plugin.getShopManager().removeShop(this);
+            onUnload();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:448]: Why we need to call `onUnload()` for double times?

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 6]
```diff
@@ -747,9 +801,10 @@ public void refresh() {
         }
         if (displayItem != null) {
             displayItem.remove();
-            initDisplayItem();
-            displayItem.spawn();
         }
+        initDisplayItem();
+        displayItem.spawn();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:806]: Why we need to put two statements to the outside?

**Review by sandtechnology — CHANGES_REQUESTED** [review]
We should as add a config option for toggling this (`true` by default), and I'm not sure if the `refersh()` method will work properly in any time, let me check that later.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 7]
```diff
@@ -1175,15 +1239,65 @@ public boolean isDoubleShop() {
...
+
+    /**
+     * This function calculates which block of a double chest is the left block,
+     * relative to the direction the chest is facing. Left shops don't spawn items since
+     * they merge items with the right shop.
+     *
+     * @return Whether or not this shop is the left half of a double chest shop.
+     */
+    @Override
+    public boolean isLeftShop() {
```
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1266]: getAttachedShop is expensive so

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 2 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:416]: The reason for it being up there is so that it save it from deletion. If I were to put it in the if statement at the end, it wouldn't work since the shop would have been deleted at that point.

File: src/main/java/org/maxgamer/quickshop/shop/DisplayItem.java (hunk 1 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/DisplayItem.java:139]: You're very right, that was a quick fix since I already had a single ternary and I just slapped another one on.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 3 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:431]: That was a remnant of my extensive checking to make sure it unloaded the duplicates I had made via other methods. Good catch!

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 4 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:439]: Also good catch. I had lots of doubled code from the beginning when I wanted to make sure that items were properly being removed and added.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 5 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:448]: Same as above, excellent work. My brain is too used to seeing the code so I can't even pick out duplicates like that haha.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 6 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:806]: Good point, in fact it's even better to not do that just in case it gets called on a left shop which shouldn't have an item. I thought I was making it more robust but that just opens up new possibilities for duplicates.,

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java]: Accomplished in recent commits! Tougher than I presumed, I really hate switch statement auto fall through

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 8]
```diff
@@ -21,6 +21,7 @@
 
 import com.lishid.openinv.OpenInv;
 import io.papermc.lib.PaperLib;
+import jdk.internal.net.http.common.Utils;
```
Comment (Ghost-chu) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:24]: not a good idea to use internal utils

**Review by Ghost-chu — CHANGES_REQUESTED** [review]

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 8 above)
Comment (sandtechnology) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:24]: It's an unused imports lol

Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:24]: unused imports

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 9]
```diff
@@ -158,10 +166,27 @@ public ContainerShop(
...
+            if (displayItem != null) {
+                displayItem.remove();
+            }
+            // Update double shop status, is left status, and the attachedShop
+            updateAttachedShop();
+            // Don't make an item for this chest if it's a left shop.
+            if (isLeftShop) {
+                if (displayItem != null) {
+                    displayItem.remove();
+                }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:178]: Duplicated code, you have remove it double times

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 2 above)
Comment (sandtechnology) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:436]: So please add a comment for it

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 10]
```diff
@@ -747,9 +816,16 @@ public void refresh() {
...
             displayItem.remove();
+        }
+
+        checkDisplay();
+
+        if (!isLeftShop) {
             initDisplayItem();
             displayItem.spawn();
+            Util.debugLog("Not left shop, respawning item!");
         }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:827]: Recommend to remove this debug log, since not a left shop case is the most cases

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 11]
```diff
@@ -788,7 +865,6 @@ public void onLoad() {
                 }
             }
         }
-        this.checkDisplay();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:791]: Why It's was being removed?

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 12]
```diff
@@ -1160,35 +1245,93 @@ public DisplayItem getDisplayItem() {
...
+     * relative to the direction the chest is facing. Left shops don't spawn items since
+     * they merge items with the right shop.
+     * It also updates the isLeftShop status of this class to reflect the changes.
+     */
+    private void updateLeftShop() {
+        if (attachedShop == null) {
+            return;
+        }
+
+        switch (((Chest) getLocation().getBlock().getBlockData()).getFacing()) {
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1295]: We should check if the shop was instanceof Chest

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 13]
```diff
@@ -1160,35 +1245,93 @@ public DisplayItem getDisplayItem() {
...
+    /**
+     * Checks to see if it is a real double without updating anything.
+     * @return If the chest is a real double chest, as in it is a double and it has the same item.
+     */
+    public boolean isRealDouble() {
+        Util.ensureThread(false);
+        if (attachedShop == null) {
+            return false;
+        }
+        return attachedShop.matches(this.getItem());
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1324]: We should check if the shop was instanceof Chest

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java [hunk 14]
```diff
@@ -93,6 +93,10 @@ public VirtualDisplayItem(@NotNull Shop shop) throws RuntimeException {
 
     private void initFakeDropItemPacket() {
 
+        if (shop.isLeftShop()) {
+            shop.getAttachedShop().refresh();
+            return;
+        }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:99]: I recommend put this code to the statement which invoking `initFakeDropItemPacket` method, since this method is doing initFakeDropItemPacket

**Review by sandtechnology — CHANGES_REQUESTED** [review]

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 8 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:24]: Good catch, I must've alt-entered accidentally and it imported that without me noticing.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 15]
```diff
@@ -1224,85 +1245,93 @@ public DisplayItem getDisplayItem() {
...
+     * Also updates the left shop status.
      */
-    @Override
-    public boolean isRealDouble() {
+    public void updateAttachedShop() {
         Util.ensureThread(false);
-        ContainerShop nextTo = this.getAttachedShop();
-        if (nextTo == null) {
-            return false;
+        Block c = Util
```
[the remark this answers is not in the document]
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1268]: I'll admit, it took me longer than it should've to realize you wanted me to rename that very undescriptive variable. I was looking up Javadocs like "There is no method called attachedChest in Spigot!" haha.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 11 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:791]: Excellent catch! I have no recollection of removing that but it was a weird source of bugs where the items would not load in correctly and I never even though to look in the onLoad method, I literally had it folded in my IDE. It's been added back, to superb effect.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 12 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1295]: If it gets past the point where it finds its attached shop, it _must_  be a Chest at that point. No reason to check again.
`.getSecondHalf()` will return null if the given block is not a Chest.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 13 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1324]: AttachedShop is always a ContainerShop or null. No need to check.

Comment (sandtechnology) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1324]: It can be any container, not only for a chest : D

Comment (sandtechnology) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1324]: Oh I got it, thanks

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java (hunk 2 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:436]: Added!

**Review by Ghost-chu — APPROVED** [review]

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java (hunk 14 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:99]: After extensive testing, turns out this whole statement is unnecessary!

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java [hunk 16]
```diff
@@ -305,6 +305,14 @@ public void spawn() {
             return;
         }
         load();
+
+        // Can't rely on the attachedShop cache to be accurate
+        // So just try it and if it fails, no biggie
+        try {
+            shop.getAttachedShop().updateAttachedShop();
+        } catch (NullPointerException ignored) {
+        }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:314]: Why we can't just avoid NullPointerException like this?
```
Shop attachedshop=shop.getAttachedShop();
if (attachedshop != null) {
attachedshop.updateAttachedShop();
}
```
Also It just in VirtualDisplayItem, what about other display item?

**Review by sandtechnology — CHANGES_REQUESTED** [review]
Have you test the edge case? (Double shop isn't loaded in the same time, it will happen will the shop is in the chunk edge)

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java (hunk 16 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:314]: I had it like that and it did not work, so I went with this approach. I don't see a problem with it, it solved my issue! In RealDisplayItem the error is not present, and I haven't done anything with ArmorStandDisplayItem since it was deprecated. I also figured now would be a good time to remove ArmorStandDisplayItem since we're redoing the whole item thing kinda.

**Comment by Starmism** [comment]
> Have you test the edge case? (Double shop isn't loaded in the same time, it will happen when the shop is in the chunk edge)

Yes! There are no changes when a shop spans a chunk border!
https://www.dropbox.com/s/4tzaq8h73sgbgoo/QR1oMPyUPs.mp4?dl=0

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java [hunk 17]
```diff
@@ -390,13 +405,11 @@ public void sendFakeItem(@NotNull Player player) {
...
-    @Override
-    public @NotNull Location getDisplayLocation() {
-        return shop.getLocation().clone().add(0.5, 1.2, 0.5);
-    }
-
     @Override
     public boolean isSpawned() {
+        if (shop.isLeftShop()) {
+            return (Objects.requireNonNull(shop.getAttachedShop().getDisplayItem())).isSpawned();
+        }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:412]: NullPointerException?

**Review by sandtechnology — CHANGES_REQUESTED** [review]

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java (hunk 17 above)
Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:412]: If it's a left shop, it must have an attached shop not equal to null.

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java (hunk 14 above)
Comment (sandtechnology) [reply src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:99]: So why it not being removed?

Comment (Starmism) [reply src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:99]: Because I am dumb and accidentally undid it before committing lmao.

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 18]
```diff
@@ -1160,35 +1241,93 @@ public DisplayItem getDisplayItem() {
...
+                isLeftShop = getLocation().getZ() > attachedShop.getLocation().getZ();
+                break;
+            case NORTH:
+                // left block has greater x value
+                isLeftShop = getLocation().getX() > attachedShop.getLocation().getX();
+                break;
+            case SOUTH:
+                // left block has a smaller x value
+                isLeftShop = getLocation().getX() < attachedShop.getLocation().getX();
+                break;
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java:1307]: Recommanded to add default branch for setting `isLeftShop` to `false`

**Review by sandtechnology — CHANGES_REQUESTED** [review]
Will approve and merge after this issue was fixed, so Please make sure it was ready for merge.

**Review by sandtechnology — APPROVED** [review]

**Comment by sandtechnology** [comment]
Thanks for contributing and your time!
